crates/ruff/src/checkers/ast/mod.rs

```diff
@@ -134,6 +135,53 @@ impl<'a> Checker<'a> {
         }
         noqa::rule_is_ignored(code, offset, self.noqa_line_for, self.locator)
     }
+
+    /// Create a [`Generator`] to generate source code based on the current AST state.
+    pub fn generator(&self) -> Generator {
+        fn quote_style(context: &Context, locator: &Locator, indexer: &Indexer) -> Option<Quote> {
+            if !context.in_f_string() {
+                return None;
+            }
+
+            let Some(expr) = context.expr() else {
+                return None;
+            };
+
+            // Find the f-string containing the current expression.
+            let start = expr.start();
+            let string_ranges = indexer.f_string_ranges();
+            let Ok(string_range_index) = string_ranges.binary_search_by(|range| {
+                if start < range.start() {
+                    std::cmp::Ordering::Greater
+                } else if range.contains(start) {
+                    std::cmp::Ordering::Equal
+                } else {
+                    std::cmp::Ordering::Less
+                }
+            }) else {
+                return None;
+            };
```

Comment:
Nit: I think it would be good to move this logic to the `Indexer`. It makes very explicit assumption about `string_ranges` and it may not be obvious to someone working on `Indexer` that some code is making these assumptions

Reply:
Good idea.